src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -6169,25 +6219,31 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
     /// So, such case has special meaning, if drop_range contains only one block it means that nothing to drop.
     /// TODO why not to add normal DROP_RANGE entry to replication queue if `replace` is true?
     MergeTreePartInfo drop_range;
-    getFakePartCoveringAllPartsInPartition(partition_id, drop_range, true);
+    std::optional<EphemeralLockInZooKeeper> delimiting_block_lock;
+    bool partition_was_empty = !getFakePartCoveringAllPartsInPartition(partition_id, drop_range, delimiting_block_lock, true);
+    if (replace && partition_was_empty)
```

Comment:
But where is the guarantee that no concurrent insert can happen here? Or in this case it's OK?

Reply:
It's ok, all blocks with numbers less than `delimiting_block_lock->getNumber()` will be dropped, and newer blocks with greater numbers will not. But we must hold `delimiting_block_lock` until creation of `DROP_RANGE`/`REPLACE_RANGE` entry in replication log to avoid merges intersecting `delimiting_block_lock->getNumber()`.